dwave/system/temperatures.py

```diff
@@ -0,0 +1,462 @@
+# Copyright 2021 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+"""Effective temperature estimators
+
+Maximum pseudo-likelihood is an efficient estimator for the temperature 
+describing a classical Boltzmann distribution P(x) = exp(-H(x)/T)/Z(T) 
+given samples from that distribution, where H(x) is the classical energy 
+function. This estimator is implemented.
+See also: https://doi.org/10.3389/fict.2016.00023 
+https://www.jstor.org/stable/25464568
+
+A temperature can also be inferred from an assumed freeze-out phenomena in 
+combination with the schedule energy scale [B(s), the energy of the problem 
+Hamiltonian]. This estimator is implemented. Relevant device-specific properties
+are published for online solvers:
+https://docs.dwavesys.com/docs/latest/doc_physical_properties.html
+
+""" 
+
+import warnings
+import numpy as np
+import dimod
+from scipy import optimize
```

Comment:
You'll need to add scipy to the `setup.py` https://github.com/dwavesystems/dwave-system/blob/cde646be9ec3f314e5fbcb9b366aa24c62eb00d7/setup.py#L35 and to the `requirements.txt` https://github.com/dwavesystems/dwave-system/blob/cde646be9ec3f314e5fbcb9b366aa24c62eb00d7/requirements.txt#L11.

The former should have a version range, probably something like `'scipy>=1.5.2,<2.0.0'` (so that we can continue to support https://github.com/dwavesystems/penaltymodel/issues/134). The latter should have a pinned version. Ideally we would test that range in CI, but for now if you could just check on your personal system that it works with the minimum and latest that will have to suffice.

Reply:
Works with 1.5.2 and latest version. Made amendments.